chain/ethereum/src/data_source.rs

```diff
@@ -242,23 +244,53 @@ impl blockchain::DataSource<Chain> for DataSource {
             errors.push(SubgraphManifestValidationError::SourceAddressRequired.into());
         };
 
-        // Validate that there are no more than one of each type of block_handler
-        let has_too_many_block_handlers = {
-            let mut non_filtered_block_handler_count = 0;
-            let mut call_filtered_block_handler_count = 0;
-            self.mapping
-                .block_handlers
-                .iter()
-                .for_each(|block_handler| {
-                    if block_handler.filter.is_none() {
-                        non_filtered_block_handler_count += 1
-                    } else {
-                        call_filtered_block_handler_count += 1
-                    }
-                });
-            non_filtered_block_handler_count > 1 || call_filtered_block_handler_count > 1
-        };
-        if has_too_many_block_handlers {
+        // Ensure that there is at most one instance of each type of block handler
+        // and that a combination of a non-filtered block handler and a filtered block handler is not allowed.
+
+        let mut non_filtered_block_handler_count = 0;
+        let mut call_filtered_block_handler_count = 0;
+        let mut polling_filtered_block_handler_count = 0;
+        let mut initialization_handler_count = 0;
+        self.mapping
+            .block_handlers
+            .iter()
+            .for_each(|block_handler| {
+                match block_handler.filter {
+                    None => non_filtered_block_handler_count += 1,
+                    Some(ref filter) => match filter {
+                        BlockHandlerFilter::Call => call_filtered_block_handler_count += 1,
+                        BlockHandlerFilter::Once => initialization_handler_count += 1,
+                        BlockHandlerFilter::Polling { every: _ } => {
+                            polling_filtered_block_handler_count += 1
+                        }
+                    },
+                };
+            });
+
+        let has_non_filtered_block_handler = non_filtered_block_handler_count > 0;
+        // If there is a non-filtered block handler, we need to check if there are any
+        // filtered block handlers
+        // If there are, we do not allow that combination
+        let has_both_filtered_and_non_filtered = has_non_filtered_block_handler
+            && (call_filtered_block_handler_count > 0
+                || polling_filtered_block_handler_count > 0
+                || initialization_handler_count > 0);
+
+        if has_both_filtered_and_non_filtered {
+            errors.push(anyhow!(
+                "data source has both filtered and non-filtered block handlers, \
+                     this is not allowed"
+            ));
+        }
```

Comment:
We currently allow having a call filter and an unfiltered block handler in the same ds. So this would be a breaking change, I think we can continue to allow that combination.

But do we really want to allow combining polling and initialization handlers?

Reply:
> We currently allow having a call filter and an unfiltered block handler in the same ds

@leoyvens do people actually use that combination? I don't see the point in having any filtered handler when there is a non-filtered block handler. is it only because its a breaking change that we want to allow that?

Reply:
If we can use some tooling to determine that no subgraph uses this combination, we could disallow it. But unless we can prove otherwise, [this usually happens](https://xkcd.com/1172/).